circuit-macros/src/circuit_type.rs

```diff
@@ -0,0 +1,368 @@
+}
+
+/// Convert an `Ident` directly into a `Path`
+fn path_from_ident(identifier: Ident) -> Path {
+    parse_quote!(#identifier)
+}
+
+/// Implements a serialization function that looks like
+///     fn #method_name(self) -> Vec<#target_type> {
+///         vec![self.field1, self.field2, ...]
+///     }
+fn build_serialize_method(
+    method_name: Ident,
+    target_type: Path,
+    self_struct: &ItemStruct,
+) -> TokenStream2 {
+    let mut field_exprs: Vec<Stmt> = Vec::with_capacity(self_struct.fields.len());
+    for field in self_struct.fields.iter().cloned() {
+        let field_ident = field.ident;
+        field_exprs.push(parse_quote! {
+            res.extend(self.#field_ident.#method_name());
+        });
+    }
+
+    let fn_impl: ItemFn = parse_quote! {
+        fn #method_name(self) -> Vec<#target_type> {
+            let mut res = Vec::new();
+            #(#field_exprs)*
+
+            res
+        }
+    };
+    fn_impl.to_token_stream()
+}
+
+/// Implements a deserialization function for a trait that looks like the following
+///     fn #method_name<I: Iterator<Item = #from_type>>(i: &mut I) -> Self {
+///         Self { field1: i.next().unwrap(), field2: , ... }
+///     }
+fn build_deserialize_method(
+    method_name: Ident,
+    from_type: Path,
+    trait_ident: Path,
+    self_struct: &ItemStruct,
+) -> TokenStream2 {
+    let mut fields_expr: Punctuated<FieldValue, Comma> = Punctuated::new();
+    for field in self_struct.fields.iter().cloned() {
+        let ident = field.ident.expect("only named fields supported");
+        let field_type = field.ty;
+
+        // The parse field expr recursively calls `#method_name` on the field type
+        let parse_field_expr: Expr = parse_quote! {
+            <#field_type as #trait_ident>::#method_name(i)
+        };
+
+        fields_expr.push(FieldValue {
+            attrs: Vec::new(),
+            member: Member::Named(ident),
+            colon_token: Some(Colon::default()),
+            expr: parse_field_expr,
+        });
+    }
+
+    parse_quote! {
+        fn #method_name<I: Iterator<Item = #from_type>>(i: &mut I) -> Self {
+            Self {
+                #fields_expr
+            }
+        }
+    }
+}
+
+/// Build a replica of the given struct with the given modifications, using an
+/// implemented trait's associated type as the new type for each field
+fn build_modified_struct_from_associated_types(
+    base_type: &ItemStruct,
+    new_name: Ident,
+    attributes: Vec<Attribute>,
+    generics: Generics,
+    type_derivation_trait_ident: Path,
+    associated_type_ident: Path,
+) -> ItemStruct {
+    // Build the fields fo the var struct
+    let new_fields = base_type
+        .fields
+        .iter()
+        .map(|f| {
+            let name = f.ident.clone();
+            let curr_type = f.ty.clone();
+
+            // Construct the fully-qualified path type expression
+            let base_trait = type_derivation_trait_ident.clone();
```

Comment:
I believe this depends on what it expands to, but it can. In terms of performance of unnecessary clones, it's not that big a deal considering this is a compile-time concern.